thunder/dynamo/compiler.py

```diff
@@ -36,27 +310,141 @@ def __init__(self, **thunder_options):
+                    return (matmul,)
+
+            # Split Graph: Since `sinc` is unsupported, we will see two partitions, one for thunder and one for inductor.
+            class GraphModule(torch.nn.Module):
+                def forward(self, l_x_: "f32[2]"):
+                    submod_1 = self.submod_1(l_x_)
+                    submod_2 = self.submod_2(l_x_, submod_1);  l_x_ = submod_1 = None
+                    return (submod_2,)
+
+                class submod_1(torch.nn.Module):  # Partition for inductor
+                    def forward(self, l_x_: "f32[2]"):
+                        y: "f32[2]" = torch.sinc(l_x_);  l_x_ = None
+                        return y
+
+                class submod_2(torch.nn.Module):  # Partition for thunder
+                    def forward(self, l_x_: "f32[2]", y: "f32[2]"):
+                        matmul: "f32[]" = torch.matmul(l_x_, y);  l_x_ = y = None
+                        return matmul
+        """
+        # Create an `ThunderOperatorSupport` instance which will be used in the callback.
+        # This will determine whether the operation represented by the node is supported by thunder.
+        operator_support = ThunderOperatorSupport(gm)
+
+        # The callback below is called for every node in the graph.
+        # It returns an `int` denoting the parition where the node should be placed.
+        # We want to partition the graph into contiguous regions (with one or more operations)
+        # into thunder supported or unsupported region.
+        # `prev_value` is used to determine if we are still in same region (i.e. supported region or unsupported region).
+        # `partition_cnt` is bumped everytime we change the region i.e. flip from supported to unsupported or from unsupported to supported.
+        # `supported_partitions` is used to track the thunder supported partitions.
+        prev_value = None
+        partition_cnt = 0
+        supported_partitions = set()
+
+        def callback(node) -> int:
+            assert node.op not in (
+                "placeholder",
+                "get_attr",
+                "output",
+            ), f"fx.split_module should have only passed node.op=call_* but received {node.op}"
+            nonlocal prev_value, partition_cnt
+            is_thunder_supported = operator_support.is_node_supported(gm, node)
+            if prev_value == is_thunder_supported:  # We are in the same region.
+                return partition_cnt
+
+            # There is a flip. Either from supported to unsupported or unsupported to supported.
+            prev_value = is_thunder_supported
+            partition_cnt += 1  # Bump the region cnt.
+
+            if is_thunder_supported:
+                supported_partitions.add(partition_cnt)
+            return partition_cnt
+
+        # `fx_split_module` iterates over nodes and determines the partition to place them based on the callback.
+        split_module: torch.fx.GraphModule = fx_split_module(
```

Comment:
Is this better than `CapabilityBasedPartitioner`? What is more used and tested in the PyTorch repo?